### PR TITLE
Multi-stage codegen process to fix ordering issues

### LINKED ISSUE #4: Ordering of code generation stages
Currently there is a circular dependency in the code generation process. Before the `controller-gen` and `angryjet` tools are run, a generated type asserted as a `resource.Managed` or `metav1.Object` will be invalid, because the methodsets that satisfy these interfaces are not yet present.

`controller-gen` and `angryjet` can't run on invalid code, so the source files that rely on these type assertions or inferences (`compare.go`, `encode.go`, `decode.go`) can't be put in place until after these tools have successfully run.

The code generation tool needs to break up the process of generating source files into a couple distinct stages for this to flow cleanly from a single execution:
-  `generated/generate.go` (which contains the `go generate` tags to run `controller-gen` and `angryjet`), and for each resource `types.go` and `doc.go`
- invoke` go generate` on the target source tree
- generate `compare.go`, `configure.go`, `decode.go`, encode.go`, `index.go`
- generate the index over all resource types (which is how main.go discovers them to be registered as controllers)

Note that all these files are currently generated, this issue is simply about refactoring the code that drives code generation to reorder them and take a break to run `go generate` after the first step.

### PR DESCRIPTION
**Pull request by kasey**
This PR fixes https://github.com/crossplane-contrib/terraform-provider-gen/issues/4

Please see https://github.com/crossplane-contrib/provider-terraform-vsphere/pull/14 for related changes in the vsphere provider to use this new codegen workflow.

In the previous workflow, the terraform-provider-gen cli would be run with an output flag pointed at the location of a generated provider's `generated/resources` tree. The plugin had to be present on the filesystem at a location specified by the user with the `--plugin-path` argument. Because the program would look for template files in the hack directory located in the terraform-provider-gen repo, the program had to be run on a filesystem with a terraform-provider-gen repo cloned to disk, either run from the root of the terraform-provider-gen repo, or with the path to the repo specified via a `--repo-root` flag.

This PR introduces significant changes to the code generation workflow. Now users do not need to learn how to directly invoke the `terraform-provider-gen` cli. Instead `go generate` is run from the generated provider repo. Assuming correct `generate.go` and `terraform-provider-gen.yaml` files, `generate.go` acts a script driving the overall codegen process, just as our go-parents always intended. In order to make this possible, the following features were added:
1) the github.com/crossplane-contrib/terraform-provider-dl is used to download the provider binary to disk. By controlling the download location, `go generate` is also able to construct the `--plugin-path` flag without user intervention.
2) A new subcommand has been added: `cmd/template-compile/main.go`. The command "compiles" the template files in `hack/templates` to a set of go packages which can be used to access the contents of the original template files. A `TemplateGetter` interface already existed in this project, which would allow template contents to be retrieved via a path, eg: `tg.Get("pkg/generator/encode.go.tmpl")`. A new `compiledTemplateGetter` implementation was added to replace the previous filesystem based getter. The `compiledTemplateGetter` preserves the `Get(<path>)` interface but retrieves the template data by by invoking a `func() string` function found via a generated dispatch table. The end result is that other repos can invoke the generator by running eg `go run github.com/crossplane-contrib/terraform-provider-gen generate types --cfg-path=terraform-provider-gen.yaml --plugin-path=tf-plugin --output-dir=generated/resources`. Note that this means that changes to the templates in `hack/templates` will need to be "compiled" from the terraform-provider-gen repo and committed back to main branch before they can be used by providers. Here is the current command to do so: `go run github.com/crossplane-contrib/terraform-provider-gen/cmd/template-compile --output=internal/template/compiled --package-root=github.com/crossplane-contrib/terraform-provider-gen/internal/template/compiled`
3) The `terraform-provider-gen generate schema` command has been replaced by 2 narrower subcommands, `generate types` (which generates `types.go` and `doc.go`) and `generate runtime`, which generates the runtime support methods (`encode.go`, `decode.go`, `compare.go`), and registers generated types with terraform-runtime (`index.go` for each type;  `generated/resources/index_resources.go` to connect the generated entrypoints to `main.go`).

This PR also adds `generator.Field` sorting in a few locations in order to achieve deterministic output, so generated code should only change based on real changes to the schema, not due to artifacts of traversing unordered go types.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -87,13 +87,20 @@ func run() error {
...
 		st := provider.NewSchemaTranslator(cfg, *outputDir, *overlayBasePath, p.GetSchema(), tg)
-		return st.WriteAllGeneratedResourceFiles()
+
+		switch cmd {
+		case generateTypesCmd.FullCommand():
+			return st.WriteGeneratedTypes()
+		case generateRuntimeCmd.FullCommand():
+			return st.WriteGeneratedRuntime()
+		}
+		//return st.WriteAllGeneratedResourceFiles()
```
Comment (hasheddan) [inline main.go]: Is this meant to be commented out?

Comment (kasey) [reply main.go]: ✂️

**Review by hasheddan — APPROVED** [review]
Thanks for tackling this @kasey!

**Comment by kasey** [comment]
@hasheddan merging this down now to hopefully fix the build over at https://github.com/crossplane-contrib/provider-terraform-vsphere/pull/14

**Pull request merged**